Option to cancel reload on weapon switch (Experimental)

pixelrider2000 opened the pull request:
This is how I've implemented reload cancellation without major rewrites.
The way it works is by saving the stack to reload in the guns nbt data instead of immediately inserting it.
A 'QueuedReload' object is created and once the reload is finished server side, the magazine is inserted.

Note:
When you have only one magazine left in your inventory and then reload an empty gun and cancel this reload. The gun appears empty to the player, though can still be reloaded since there is still a magazine queued. 

Perhaps there should be a way of indicating that there is still an unfinished reload. 
Maybe by making the ammo indicator next to your hotbar a little transparent?

Just some thoughts. Would love to hear some feedback!

pixelrider2000 committed:
D

pixelrider2000 committed:
Delete gradlew.bat

pixelrider2000 committed:
Fix possible issue with mechas

GoldSloth commented on src/main/java/com/flansmod/common/PlayerData.java:
```diff
@@ -114,6 +121,29 @@ public void tick(EntityPlayer player)
+			
+			for(QueuedReload qR : new ArrayList<>(queuedReloads)) {
+				if(qR.getReloadTime() > 0) {
+					qR.decrementReloadTime();
```
Does this mean you can have more than one gun reloading at once, somehow?

pixelrider2000 replied:
No. Some guns have more than one magazine (shotguns). Each itemstack is handled individually. Now that I think about it though, there is no reason it needs to be that way. I'm gonna change that probably...

GoldSloth replied:
ah that makes sense.

GoldSloth commented on src/main/java/com/flansmod/common/guns/ItemGun.java:
```diff
@@ -1312,29 +1325,46 @@ public boolean reload(ItemStack gunStack, GunType gunType, World world, Entity e
+                	if(!FlansMod.cancelReloadOnWeaponSwitch || !isPlayer) {
+	                    //Unload the old magazine (Drop an item if it is required and the player is not in creative mode)
+	                    if (bulletStack != null && bulletStack.getItem() instanceof ItemShootable && ((ItemShootable) bulletStack.getItem()).type.dropItemOnReload != null && !creative && bulletStack.getItemDamage() == bulletStack.getMaxDamage())
+	                        dropItem(world, entity, ((ItemShootable) bulletStack.getItem()).type.dropItemOnReload);
```
would this stop dropItemOnReload working if you had it enabled?

GoldSloth commented on src/main/java/com/flansmod/common/guns/QueuedReload.java:
```diff
@@ -0,0 +1,74 @@
+		if(bulletInGun != null) {
+			//Unload the old magazine (Drop an item if it is required and the player is not in creative mode)
+            if (bulletInGun != null && bulletInGun.getItem() instanceof ItemShootable && ((ItemShootable) bulletInGun.getItem()).type.dropItemOnReload != null && !creative && bulletInGun.getItemDamage() == bulletInGun.getMaxDamage())
+                gun.dropItem(world, entity, ((ItemShootable) bulletInGun.getItem()).type.dropItemOnReload);
```
nvm, dropItemOnReload then happens here instead

GoldSloth reviewed (commented):
It looks okay to me, but I'd like to test it tomorrow a bit. @Unknown025 I'd appreciate if you could check this too as it's the sort of change that could introduce unwanted behaviour regarding inventory e.t.c.

GoldSloth commented:
Sorry, not sure what you mean by this:

`When you have only one magazine left in your inventory and then reload an empty gun and cancel this reload. The gun appears empty to the player, though can still be reloaded since there is still a magazine queued.`

Also, just wanted to check that the magazine gets taken out of the inventory only after the reload finishes - ie if the player takes the mag out of the inventory somehow, it cancels it before the mag is placed into the gun.

pixelrider2000 commented:
Well, the magazine is still taken out of the players inventory as soon as you press R to reload. The gun now holds an "active" magazine and a "queued" magazine. Let's say you cancelled the reload, the magazine is still queued and "inside the gun". It would be possible to give it back to the player once the reload is cancelled, I suppose. 

That means there is no way for the player to drop the magazine before it is placed into the gun, because he doesn't have it in his inventory anymore when reloading. That's btw. precisely why I chose to do it that way.

Better idea?:
Maybe it would be smarter to simply check if the player can reload when 'R' is pressed, but perform the actual reload only when the reload actually finished. 
If the player had dropped the magazine by then, the gun would simply not be loaded, since it does all the checking again.

I think this can be well improved still. I see if I can find a better solution.

pixelrider2000 committed:
Improved reload cancel

pixelrider2000 commented:
I have now massively changed how everything works. I do believe it is much better now.
This is what it looks like at the moment: https://streamable.com/8y7rlj

brandyyn commented:
> I have now massively changed how everything works. I do believe it is much better now. This is what it looks like at the moment: https://streamable.com/8y7rlj

LETS FUCKING GOOOOOO MY GOAT

GoldSloth commented:
Ahhh I see what you mean about it not showing in the GUI. It would be good to show both, maybe have the reloading one flash or something like that. So you can see the ammunition that's currently being loaded, as well as what's still in the gun. They could fade in/out as reload animation progresses too perhaps?

GoldSloth commented on src/main/java/com/flansmod/client/model/GunAnimations.java:
```diff
@@ -195,7 +206,24 @@ public void update() {
         //Reload
+        if(FlansMod.cancelReloadOnWeaponSwitch) {
+            ItemStack heldItem = Minecraft.getMinecraft().thePlayer.getHeldItem();
+            if(reloading && gunToReload != null && (heldItem == null || !gunToReload.getItem().equals(heldItem.getItem()))) {
```
Doing this in gunanimations seems a bit dodgy, ideally GunAnimations only holds stuff that's visual - shootTimeRightLeft e.t.c.
TickHandlerClient might be better? Or FlansModClient if it has a regular update thingy... I think this sort of thing is probably in there, or even ItemGun itself.

Also, with setting the shoot time here, would that mean you can cancel a really long shoot delay, e.g. shoot once, reload, cancel the reload and the shoot time is now zero? Putting in some kind of penalty for that might be needed.

GoldSloth replied:
Just realised this isn't an issue because you can't start a reload before the shoot time reaches zero anyway..

pixelrider2000 commented:
I'm not sure if it's really worth the effort. Games like Battlefield also don't show the ammunition that's being loaded.
I think the fact that you can reload the gun tells the player all they need to know - and that is that they still have ammo.

And besides, having the counter do one thing - and that is show what's in the gun RIGHT NOW - might even make things easier to read. And how would you deal with guns that can load multiple stacks? 
I believe it's just a bit too much at the moment (though one could add that later).

GoldSloth commented:
Yeah that's fair enough. Thought it just looked a bit odd, especially when changing ammunition type.

GoldSloth commented:
Think I might have found a bug, when reloading a gun then switching to a gun of the same type, the reload completes regardless. I have a feeling that's because the item stack comparison still passes despite them being different, because they're still the same item.

https://streamable.com/84i3z2

GoldSloth reviewed (changes requested):
Looks good, but a few changes needed I think (and the bug I mentioned in the main thread). Overall really good work though :))

Let me know when you want me to do some more testing btw.

brandyyn commented:
> Think I might have found a bug, when reloading a gun then switching to a gun of the same type, the reload completes regardless. I have a feeling that's because the item stack comparison still passes despite them being different, because they're still the same item.
> 
> https://streamable.com/84i3z2

can check for slot change rather than item change

pixelrider2000 committed:
Improve reload cancel and bug fix

pixelrider2000 commented:
Alright, I cleaned up the code a bit. Hopefully I fixed the bug too, although I'm not sure if you should compare item stacks like that in forge. It worked for my testing, but idk, it seems a bit weird to compare their nbt in that way. Is there a better way?

pixelrider2000 commented:
> > Think I might have found a bug, when reloading a gun then switching to a gun of the same type, the reload completes regardless. I have a feeling that's because the item stack comparison still passes despite them being different, because they're still the same item.
> > https://streamable.com/84i3z2
> 
> can check for slot change rather than item change

The reason this doesn't work is because of dropping the weapon or your inventory getting cleared or else.

GoldSloth commented on src/main/java/com/flansmod/client/model/GunAnimations.java:
```diff
@@ -5,8 +5,17 @@
+import com.flansmod.common.PlayerHandler;
 import com.flansmod.common.vector.Vector3f;
 
+import cpw.mods.fml.relauncher.Side;
```
Much better :)

Though these imports aren't needed now

GoldSloth commented on src/main/java/com/flansmod/common/PlayerData.java:
```diff
@@ -114,6 +121,38 @@ public void tick(EntityPlayer player)
+		        	gunAnimationLeft.reloading = false;
+				} 
+			} else if(!player.worldObj.isRemote && queuedReload != null) {
+				if(queuedReload.getReloadTime() > 0) {
```
does the formatting look normal in your IDE? Idk if it's just Github that's made it look really weird

Unknown025 replied:
Formatting looks off to me as well.

GoldSloth commented on src/main/java/com/flansmod/common/PlayerData.java:
```diff
@@ -131,6 +170,18 @@ public void tick(EntityPlayer player)
+	public boolean isHoldingGunToReload(EntityPlayer player) {
+    	ItemStack heldItem = player.getHeldItem();
+
+    	return !(heldItem == null || gunToReload.stackTagCompound != heldItem.stackTagCompound);
```
Hm... still not sure this will do the full equality. Idk if there's some unique item ID?
If this works, what's the unique thing in their NBT, other than current ammo state?

I think different slot may genuinely work quite well, if someone's inv is wiped or the item is somehow removed, the reload cancelling isn't the end of the world. Still not 100% sure.. there must be a better way?

Unknown025 replied:
You could maybe compare the item directly, since their memory addresses should match.

GoldSloth commented:
Looks good :) As mentioned not sure about the comparison still, but I did some testing and it seems to be working alright?

I'd still like @Unknown025 to review it before merging, but it's not far off now at all.

Unknown025 reviewed (approved):
Looks good overall. If the code formatting could be fixed up a little, I'd appreciate it. Otherwise, no other problems stand out to me. The system may not be perfect, but as long as the normal behavior can be enabled, I don't have a problem with it.

pixelrider2000 committed:
Unused imports and gun comparison

pixelrider2000 commented:
I now compare the item stacks directly. I swear that I've done that before and it didn't seem to work for some reason. That's why I didn't even attempt it this time, but well, it seems to be working the same after all... magic

pixelrider2000 commented:
This is what the formatting looks like in my IDE. All good I would say:
![Screenshot 2024-02-17 092701](https://github.com/Unknown025/Flans-Mod-Plus/assets/76816145/07da9f1d-4427-428e-9862-c179dc4ee189)

GoldSloth commented:
Is your IDE set to use tabs or spaces? That might be causing it perhaps?

pixelrider2000 committed:
Fix formatting

pixelrider2000 commented:
Better? :D
It was set to tabs btw.

GoldSloth commented:
That looks better, I'm happy with that. Thanks :))

brandyyn commented:
Yayyyyyyyyy my goats

brandyyn commented:
not sure if this is caused by this pr? https://streamable.com/svghck

pixelrider2000 commented:
> not sure if this is caused by this pr? https://streamable.com/svghck

Definitely is caused by this. I think I already know what's causing it, but I'm not 100% sure what the best way to fix this is. 
I've posted my first idea on discord though. Would like to get some other ideas in if possible.
